python/ray/util/collective/collective.py

```diff
@@ -0,0 +1,261 @@
+"""APIs exposed under the namespace ray.util.collective."""
+import logging
+
+import ray
+from ray.util.collective import types
+from ray.util.collective.const import get_nccl_store_name
+
+# Get the availability information first by importing information
+_MPI_AVAILABLE = False
+_NCCL_AVAILABLE = True
+
+# try:
+#     from ray.util.collective.collective_group.mpi_collective_group
+#     import MPIGroup
+# except ImportError:
+#     _MPI_AVAILABLE = False
+
+try:
+    from ray.util.collective.collective_group import NCCLGroup
+except ImportError:
+    _NCCL_AVAILABLE = False
+
+logger = logging.getLogger(__name__)
+
+
+def nccl_available():
+    return _NCCL_AVAILABLE
+
+
+def mpi_available():
+    return _MPI_AVAILABLE
+
+
+class GroupManager(object):
+    """
+    Use this class to manage the collective groups we created so far.
+
+    """
+
+    def __init__(self):
+        """Put some necessary meta information here."""
+        self._name_group_map = {}
+        self._group_name_map = {}
+
+    def create_collective_group(self, backend, world_size, rank, group_name):
+        """
+        The entry to create new collective groups and register in the manager.
+
+        Put the registration and the group information into the manager
+        metadata as well.
+        """
+        backend = types.Backend(backend)
+        if backend == types.Backend.MPI:
+            raise NotImplementedError()
+        elif backend == types.Backend.NCCL:
+            # create the ncclUniqueID
+            if rank == 0:
+                import cupy.cuda.nccl as nccl
+                group_uid = nccl.get_unique_id()
+                store_name = get_nccl_store_name(group_name)
+                # Avoid a potential circular dependency in ray/actor.py
+                from ray.util.collective.util import NCCLUniqueIDStore
+                store = NCCLUniqueIDStore.options(
+                    name=store_name, lifetime="detached").remote(store_name)
```

Comment:
hmm, why can't the NCCL uniqueidstore just sit on the rank0 worker process?

Reply:
ah I guess you need to access it in a symmetric fashion (without communication with some overarching coordinator.

I guess maybe in the future, `actor.options(group` would make this possible.

Reply:
Yep. For now, the user's calls are supposed to be made inside the actor so one cannot touch the driver code. Expect to see something you described in the next PR.